File: go/2015/day07b/day07b.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strconv"
)

type Operator int

const (
	Assign Operator = iota
	Not
	And
	Or
	LeftShift
	RightShift
)

type Operation struct {
	operator Operator
	source1 string
	source2 string
	amount int
}

func usage(progname string) {
	fmt.Fprintf(os.Stderr, "usage: %s <input file>\n", progname)
	os.Exit(1)
}

func process(filename string) int {
	operations := make(map[string]Operation)
	cache := make(map[string]int)
	re1 := regexp.MustCompile(`^(\d+|\w+) -> (\w+)$`)
	re2 := regexp.MustCompile(`NOT (\d+|\w+) -> (\w+)$`)
	re3 := regexp.MustCompile(`(\d+|\w+) (AND|OR) (\d+|\w+) -> (\w+)`)
	re4 := regexp.MustCompile(`(\d+|\w+) (LSHIFT|RSHIFT) (\d+) -> (\w+)`)

	file, err := os.Open(filename)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error opening file: %v\n", err)
		os.Exit(1)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		matches := re1.FindStringSubmatch(line)
		if matches != nil {
			src := matches[1]
			dest := matches[2]
			operations[dest] = Operation{ Assign, src, "", 0 }
		}
		matches = re2.FindStringSubmatch(line)
		if matches != nil {
			src := matches[1]
			dest := matches[2]
			operations[dest] = Operation{ Not, src, "", 0 }
		}
		matches = re3.FindStringSubmatch(line)
		if matches != nil {
			src1 := matches[1]
			op := matches[2]
			src2 := matches[3]
			dest := matches[4]
			if op == "AND" {
				operations[dest] = Operation{ And, src1, src2, 0 }
			} else {
				operations[dest] = Operation{ Or, src1, src2, 0 }
			}
		}
		matches = re4.FindStringSubmatch(line)
		if matches != nil {
			src := matches[1]
			op := matches[2]
			amt, _ := strconv.Atoi(matches[3])
			dest := matches[4]
			if op == "LSHIFT" {
				operations[dest] = Operation{ LeftShift, src, "", amt }
			} else {
				operations[dest] = Operation{ RightShift, src, "", amt }
			}
		}
	}

	if err := scanner.Err(); err != nil {
		fmt.Fprintf(os.Stderr, "Error reading file: %v\n", err)
		os.Exit(1)
	}
	a := eval(operations, cache, "a")
	operations["b"] = Operation{ Assign, strconv.Itoa(a), "", 0 }
	cache = make(map[string]int)
	return eval(operations, cache, "a")
}

func eval(ops map[string]Operation, cache map[string]int, expr string) int {
	n1, err := strconv.Atoi(expr)
	if err == nil {
		return n1
	}
	n2, exists := cache[expr]
	if exists {
		return n2
	}
	op := ops[expr]
	r := 0
	switch op.operator {
	case Assign:
		src := op.source1
		a := eval(ops, cache, src)
		r = a
	case Not:
		src := op.source1
		a := eval(ops, cache, src)
		r = ^a
	case And:
		src1 := op.source1
		src2 := op.source2
		a := eval(ops, cache, src1)
		b := eval(ops, cache, src2)
		r = a & b
	case Or:
		src1 := op.source1
		src2 := op.source2
		a := eval(ops, cache, src1)
		b := eval(ops, cache, src2)
		r = a | b
	case LeftShift:
		src := op.source1
		amt := op.amount
		a := eval(ops, cache, src)
		r = a << amt
	case RightShift:
		src := op.source1
		amt := op.amount
		a := eval(ops, cache, src)
		r = a >> amt
	}
	masked := r & 0xffff
	cache[expr] = masked
	return masked
}

func main() {
	if len(os.Args) < 2 {
		usage(os.Args[0])
	}

	filename := os.Args[1]
	result := process(filename)
	fmt.Printf("result = %d\n", result)
}

File: dart/2015/day07b/day07b.dart
#!/usr/bin/env dart

import 'dart:collection';
import 'dart:io';

sealed class Operation {}

class Assign extends Operation {
  final String src;
  Assign(this.src);
}

class Not extends Operation {
  final String src;
  Not(this.src);
}

class And extends Operation {
  final String src1;
  final String src2;
  And(this.src1, this.src2);
}

class Or extends Operation {
  final String src1;
  final String src2;
  Or(this.src1, this.src2);
}

class LeftShift extends Operation {
  final String src;
  final int amt;
  LeftShift(this.src, this.amt);
}

class RightShift extends Operation {
  final String src;
  final int amt;
  RightShift(this.src, this.amt);
}

void usage(String progName) {
    print('usage: $progName <input file>');
    exit(1);
}

int process(String filename) {
    final HashMap<String, Operation> operations = HashMap();
    HashMap<String, int> cache = HashMap();
    final lines = File(filename).readAsLinesSync();
    for ( final line in lines ) {
        final re1 = RegExp(r'^(\d+|\w+) -> (\w+)$');
        var match = re1.firstMatch(line);
        if ( match != null ) {
            final src = match.group(1)!;
            final dest = match.group(2)!;
            operations[dest] = Assign(src);
        }
        final re2 = RegExp(r'NOT (\d+|\w+) -> (\w+)');
        match = re2.firstMatch(line);
        if ( match != null ) {
            final src = match.group(1)!;
            final dest = match.group(2)!;
            operations[dest] = Not(src);
        }
        final re3 = RegExp(r'(\d+|\w+) (AND|OR) (\d+|\w+) -> (\w+)');
        match = re3.firstMatch(line);
        if ( match != null ) {
            final src1 = match.group(1)!;
            final op = match.group(2)!;
            final src2 = match.group(3)!;
            final dest = match.group(4)!;
            if ( op == 'AND' ) {
                operations[dest] = And(src1, src2);
            } else {
                operations[dest] = Or(src1, src2);
            }
        }
        final re4 = RegExp(r'(\d+|\w+) (LSHIFT|RSHIFT) (\d+) -> (\w+)');
        match = re4.firstMatch(line);
        if ( match != null ) {
            final src = match.group(1)!;
            final op = match.group(2)!;
            final amt = int.parse(match.group(3)!);
            final dest = match.group(4)!;
            if ( op == 'LSHIFT' ) {
                operations[dest] = LeftShift(src, amt);
            } else {
                operations[dest] = RightShift(src, amt);
            }
        }
    }
    final a = eval(operations, cache, 'a');
    operations['b'] = Assign(a.toString());
    cache = HashMap();
    return eval(operations, cache, 'a');
}

int eval(HashMap<String, Operation> ops, HashMap<String, int> cache, String expr) {
    if ( RegExp(r'^\d+$').hasMatch(expr) ) {
        return int.parse(expr);
    } else if ( cache.containsKey(expr) ) {
        return cache[expr]!;
    } else {
        final op = ops[expr]!;
        final r = switch ( op ) {
            Assign(:var src) => eval(ops, cache, src),
            Not(:var src) => ~eval(ops, cache, src),
            And(:var src1, :var src2) => eval(ops, cache, src1) & eval(ops, cache, src2),
            Or(:var src1, :var src2) => eval(ops, cache, src1) | eval(ops, cache, src2),
            LeftShift(:var src, :var amt) => eval(ops, cache, src) << amt,
            RightShift(:var src, :var amt) => eval(ops, cache, src) >> amt,
        };
        final masked = r & 0xffff;
        cache[expr] = masked;
        return masked;
    }
}

void main(List<String> args) {
    if ( args.isEmpty ) {
        usage(Platform.script.pathSegments.last);
    }
    final filename = args[0];
    final result = process(filename);
    print('result = $result');
}
